text/0000-api-evolution.md

```diff
@@ -0,0 +1,798 @@
+enough (and can usually be written in other ways) that this breakage is
+considered minor. (However, one should take into account how likely it is that
+the function in question is being called with explicit type arguments).  This
+RFC also suggests adding a `...` notation to explicit parameter lists to keep
+them open-ended (see suggested language changes).
+
+Such changes are an important ingredient of abstracting to use generics, as
+described next.
+
+#### Minor change: generalizing to generics.
+
+The type of an argument to a function, or its return value, can be *generalized*
+to use generics, including by introducing a new type parameter (as long as it
+can be instantiated to the original type). For example, the following change is
+allowed:
+
+```rust
+// MINOR CHANGE
+
+// Before
+fn foo(x: u8) -> u8;
+fn bar<T: Iterator<Item = u8>>(t: T);
+
+// After
+fn foo<T: Add>(x: T) -> T;
+fn bar<T: IntoIterator<Item = u8>>(t: T);
+```
+
+because all existing uses are instantiations of the new signature. On the other
+hand, the following isn't allowed in a minor revision:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+fn foo(x: Vec<u8>);
+
+// After
+fn foo<T: Copy + IntoIterator<Item = u8>>(x: T);
+```
+
+because the generics include a constraint not satisfied by the original type.
+
+Introducing generics in this way can potentially create type inference failures,
+but these are considered acceptable per the principles of the RFC: they only
+require local annotations that could have been inserted in advance.
+
+Perhaps somewhat surprisingly, generalization applies to trait objects as well,
+given that every trait implements itself:
+
+```rust
+// MINOR CHANGE
+
+// Before
+fn foo(t: &Trait);
+
+// After
+fn foo<T: Trait + ?Sized>(t: &T);
+```
+
+(The use of `?Sized` is essential; otherwise you couldn't recover the original
+signature).
+
+### Lints
+
+#### Minor change: introducing new lint warnings/errors
+
+Lints are considered advisory, and changes that cause downstream code to receive
+additional lint warnings/errors are still considered "minor" changes.
+
+Making this work well in practice will likely require some infrastructure work
+along the lines of
+[this RFC issue](https://github.com/rust-lang/rfcs/issues/1029)
+
+## Mitigation for minor changes
+
+### The Crater tool
+
+@brson has been hard at work on a tool called "Crater" which can be used to
+exercise changes on the entire crates.io ecosystem, looking for
+regressions. This tool will be indispensable when weighing the costs of a minor
+change that might cause some breakage -- we can actually gauge what the breakage
+would look like in practice.
+
+While this would, of course, miss code not available publicly, the hope is that
+code on crates.io is a broadly representative sample, good enough to turn up
+problems.
+
+Any breaking, but minor change to the standard library must be evaluated through
+Crater before being committed.
```

Comment:
I don't think we want to commit to running Crater with *every* minor change. Is the current cadence of once-a-week reports too slow?